src/libstore/store-api.hh

```diff
@@ -115,7 +115,8 @@ struct ValidPathInfo
 {
     StorePath path;
     std::optional<StorePath> deriver;
-    Hash narHash;
+    // TODO document this
+    std::optional<Hash> narHash;
```

Comment:
Here is one of the now-optional hashes

Reply:
@edolstra So I checked and while this one is optional, the constructor for NarInfo (derived class) requires that it is present with assertions. Is that good enough for now, or do you also want the refactor so that it is required everywhere? (I do recall you were saying the nix installer trick where the nar hash would be missing is long gone.)

(I must admit due to #3640 I am a bit partial to keeping it optional, but I'm happy to do whatever you like.)

Reply:
I think this one should *not* be optional since a store path should always have a NAR hash.

Reply:
This is actually turns out to be a fairly substantial change, because it looks like the `BinaryCacheStore::addToStore` overload that doesn't take the `ValidPathInfo` was uploading flat files as-is and calling them nars, with the `narHash` unset.

I think that a bug not intended feature, but either way, fixing it is the sort of major behavior change I rather do separately from this refactor which has no behavior changes?